### PR TITLE
Add implementation for z_sendmanywithchangetosender, which will send PSL to an address and return any change to the sending address

### PR DESCRIPTION
**Pull request by vadymeng0**
## What is this about?
Add implementation for z_sendmanywithchangetosender, which will send PSL to an address and return any change to the sending address

## What is the test results?
Built successfully
Passed all below test suite on local machine (**_in old PR #51, rerunning to validate again_**):
  cd qa/test-suite
  ./mn_main.py
  ./mn_payment.py
  ./mn_tickets.py
  ./mn_tickets_validation.py
  ./mn_messaging.py
  ./full_test_suite.py btest
  ./full_test_suite.py gtest
  ./full_test_suite.py sec-hard
  ./full_test_suite.py no-dot-so
  ./full_test_suite.py util-test
  ./full_test_suite.py secp256k1
  ./full_test_suite.py libsnark
  ./full_test_suite.py univalue
  ./full_test_suite.py rpc-ext
  ./full_test_suite.py rpc-common
  Test case that is failing:
  ./mn_governance.py <-- Failed on master branch also.
  Test case that took too long to complete, and failed some test cases while running:
  ./full_test_suite.py rpc-mn

### CODE REVIEW

File: qa/rpc-tests/wallet_sapling.py [hunk 1]
```diff
@@ -201,5 +201,63 @@ def run_test(self):
...
+
+        senderExpectedBalance = Decimal('100.0')
+        mytxid = self.nodes[0].sendtoaddress(senderaddr, senderExpectedBalance)
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+        assert_equal(self.nodes[0].z_getbalance(senderaddr), senderExpectedBalance)
+
+        receiveraddr = self.nodes[2].getnewaddress()
+        receiveraddr = self.nodes[2].getnewaddress()
```
Comment (VitaliyVK777) [inline qa/rpc-tests/wallet_sapling.py:219]: why getting receiveraddr is repeated 2 times

**Review by VitaliyVK777 — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/wallet_sapling.py (hunk 1 above)
Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py]: fixed

File: qa/rpc-tests/wallet_sapling.py [hunk 2]
```diff
@@ -201,5 +201,63 @@ def run_test(self):
         except JSONRPCException as e:
             assert_equal("Cannot send to both Sprout and Sapling addresses using z_sendmany", e.error['message'])
 
+        # Check z_sendmanywithchangetosender
+        # send from node 0 taddr to node 2 single taddr
+        senderaddr = self.nodes[0].getnewaddress()
+        senderaddr = self.nodes[0].getnewaddress()
```
[the remark this answers is not in the document]
Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py]: fixed

**Review by VitaliyVK777 — DISMISSED** [review]

File: qa/rpc-tests/wallet_sapling.py [hunk 3]
```diff
@@ -204,36 +205,44 @@ def run_test(self):
         # Check z_sendmanywithchangetosender
         # send from node 0 taddr to node 2 single taddr
         senderaddr = self.nodes[0].getnewaddress()
```
Comment (VitaliyVK777) [inline qa/rpc-tests/wallet_sapling.py:207]: I just can not get idea of changes. probably should be next?
senderaddr = self.nodes[2].getnewaddress()

**Review by VitaliyVK777 — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/wallet_sapling.py (hunk 3 above)
Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py:207]: what do you mean? And what changes you are requesting?

Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py:207]:
```
        # Check z_sendmanywithchangetosender
        # send from node 0 taddr to node 2 single taddr
```

Comment (Dicklesworthstone) [reply qa/rpc-tests/wallet_sapling.py:207]: Yes this doesn’t seem like a helpful comment…

File: qa/rpc-tests/wallet_sapling.py [hunk 4]
```diff
@@ -201,5 +202,68 @@ def run_test(self):
...
+        self.sync_all()
+        assert_equal(self.nodes[2].z_getbalance(receiveraddr), receiverExpectedBalance)
+        calculatedSenderBalance = senderExpectedBalance - Decimal('5.0') - fee
+        assert_equal(self.nodes[0].z_getbalance(senderaddr), calculatedSenderBalance)
+
+        # send from node 0 taddr to node 2 multiple taddr
+        receiveraddr = self.nodes[2].getnewaddress()
+        receiveraddr_oldbalance = self.nodes[2].z_getbalance(receiveraddr)
+        receiveraddr2 = self.nodes[2].getnewaddress()
+        receiveraddr2_oldbalance = self.nodes[2].z_getbalance(receiveraddr)
```
Comment (VitaliyVK777) [inline qa/rpc-tests/wallet_sapling.py:238]: receiveraddr2_oldbalance = self.nodes[2].z_getbalance(receiveraddr2)

File: qa/rpc-tests/wallet_sapling.py (hunk 3 above)
Comment (VitaliyVK777) [reply qa/rpc-tests/wallet_sapling.py:207]: It is clear now,

**Review by VitaliyVK777 — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/wallet_sapling.py (hunk 4 above)
Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py]: fixed

File: qa/rpc-tests/wallet_sapling.py [hunk 5]
```diff
@@ -201,5 +202,68 @@ def run_test(self):
...
+
+        recipients  = [ {"address": receiveraddr, "amount": Decimal('5.0')},
+                        {"address": receiveraddr2, "amount": Decimal('5.0')}  ]
+
+        myopid = self.nodes[0].z_sendmanywithchangetosender(senderaddr, recipients, minconf, fee)
+        mytxid = wait_and_assert_operationid_status(self.nodes[0], myopid)
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+        calculatedSenderBalance = calculatedSenderBalance - fee - receiverExpectedBalance * 2
```
Comment (VitaliyVK777) [inline qa/rpc-tests/wallet_sapling.py:251]: should be here correct is next?
        calculatedSenderBalance = calculatedSenderBalance - fee - 5.0 * 2
as 'receiverExpectedBalance' includes 'receiveraddr_oldbalance' that can be not 0

**Review by VitaliyVK777 — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/wallet_sapling.py (hunk 5 above)
Comment (vadymeng0) [reply qa/rpc-tests/wallet_sapling.py]: thanks. fixed.

**Review by VitaliyVK777 — APPROVED** [review]

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 6]
```diff
@@ -65,8 +65,9 @@ AsyncRPCOperation_sendmany::AsyncRPCOperation_sendmany(
         std::vector<SendManyRecipient> zOutputs,
         int minDepth,
         CAmount fee,
-        UniValue contextInfo) :
-        tx_(contextualTx), fromaddress_(fromAddress), t_outputs_(tOutputs), z_outputs_(zOutputs), mindepth_(minDepth), fee_(fee), contextinfo_(contextInfo)
+        UniValue contextInfo,
+        bool returnChangeToSenderAddr) :
```
Comment (akobrin1) [inline src/wallet/asyncrpcoperation_sendmany.cpp:69]: Just FYI: C++ now allows to use in constructor variables with same names as class members.
Also, can use "const bool bReturnChangeToSenderAddr" here.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -4251,12 +4253,21 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
 
     // Create operation and add to global queue
     std::shared_ptr<AsyncRPCQueue> q = getAsyncRPCQueue();
-    std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_sendmany(builder, contextualTx, fromaddress, taddrRecipients, zaddrRecipients, nMinDepth, nFee, contextInfo) );
+    std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_sendmany(builder, contextualTx, fromaddress, taddrRecipients, zaddrRecipients, nMinDepth, nFee, contextInfo, returnChangeToSenderAddr) );
```
Comment (akobrin1) [inline src/wallet/rpcwallet.cpp:4256]: it is recommended to use std::make_shared to create shared_ptr - it allocates memory in one shot for both shared_ptr control block and the object.
auto operation = std::make+shared<AsyncRPCOperation_sendmany>(builder, contextualTx, fromaddress, taddrRecipients, zaddrRecipients, nMinDepth, nFee, contextInfo, returnChangeToSenderAddr);

**Review by akobrin1 — APPROVED** [review]
Nothing critical. If you will make any changes here - you can correct shared_ptr -> make_shared

File: src/wallet/rpcwallet.cpp (hunk 7 above)
Comment (vadymeng0) [reply src/wallet/rpcwallet.cpp]: fixed as suggestion

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 6 above)
Comment (vadymeng0) [reply src/wallet/asyncrpcoperation_sendmany.cpp]: fixed as suggestion.

**Review by VitaliyVK777 — APPROVED** [review]

**Review by a-ok123 — APPROVED** [review]

**Pull request merged**